### PR TITLE
[CELEBORN-832] Support use RESTful API to trigger worker decommission

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
As title

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 1]
```diff
@@ -2469,6 +2475,23 @@ object CelebornConf extends Logging {
...
+    buildConf("celeborn.worker.decommission.check.interval")
+      .categories("worker")
+      .doc(
+        "The wait interval of checking whether all the shuffle expired during worker decomission")
+      .version("0.2.0")
+      .timeConf(TimeUnit.MILLISECONDS)
+      .createWithDefaultString("30s")
+
+  val WORKER_DECOMMISSION_CHECK_TIMEOUT: ConfigEntry[Long] =
+    buildConf("celeborn.worker.decommission.check.timeout")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2488]:
```suggestion
    buildConf("celeborn.worker.decommission.forceExitTimeout")
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -2469,6 +2475,23 @@ object CelebornConf extends Logging {
       .timeConf(TimeUnit.MILLISECONDS)
       .createWithDefaultString("10min")
 
+  val WORKER_DECOMMISSION_CHECK_INTERVAL: ConfigEntry[Long] =
+    buildConf("celeborn.worker.decommission.check.interval")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2479]:
```suggestion
    buildConf("celeborn.worker.decommission.checkInterval")
```

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 3]
```diff
@@ -592,7 +603,42 @@ private[celeborn] class Worker(
       logWarning(s"Waiting for all PartitionLocation release cost ${waitTime}ms, " +
         s"unreleased PartitionLocation: \n$partitionLocationInfo")
     }
-    stop(CelebornExitKind.WORKER_GRACEFUL_SHUTDOWN)
+  }
+
+  def decommissionWorker(): Unit = {
+    try {
+      masterClient.askSync[PbWorkerLostResponse](
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:610]: Report WorkerLost is useless while heartbeat still report worker alive to master, probably we need a new decommission(or may use shutdown status, I think it's ok) status to represent current worker state.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 4]
```diff
@@ -2469,6 +2475,23 @@ object CelebornConf extends Logging {
...
+      .categories("worker")
+      .doc(
+        "The wait interval of checking whether all the shuffle expired during worker decomission")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2483]:
```suggestion
      .version("0.4.0")
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 5]
```diff
@@ -2469,6 +2475,23 @@ object CelebornConf extends Logging {
...
+      .timeConf(TimeUnit.MILLISECONDS)
+      .createWithDefaultString("30s")
+
+  val WORKER_DECOMMISSION_CHECK_TIMEOUT: ConfigEntry[Long] =
+    buildConf("celeborn.worker.decommission.check.timeout")
+      .categories("worker")
+      .doc("The wait time of waiting for all the shuffle expire during worker decommission.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2491]:
```suggestion
      .version("0.4.0")
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 6]
```diff
@@ -2469,6 +2475,23 @@ object CelebornConf extends Logging {
...
+      .timeConf(TimeUnit.MILLISECONDS)
+      .createWithDefaultString("3d")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2493]: Maybe too long? 2h should be enough for small/middle scale

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 7]
```diff
@@ -561,6 +561,17 @@ private[celeborn] class Worker(
     sb.toString()
   }
 
+  override def decommission: String = {
+    exitKind = CelebornExitKind.WORKER_DECOMMISSION
+    new Thread() {
+      override def run(): Unit = {
+        Thread.sleep(30000)
```
Comment (pan3793) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:568]: Emm.. why should defer to trigger the exit?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: DOne

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:610]: I have checked that if we sent WorkerLost, this worker will be removed from `statusSystem.workers`, then when heartbeat, it won't be removed from `excludedWorkers`, only will be added back after a `RegisterWorker`

Comment (AngersZhuuuu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala]: Yea, should also use shutdown status, it's same meaning

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2493]: In our prod, it's always 2d...

File: service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala [hunk 8]
```diff
@@ -67,6 +67,8 @@ abstract class HttpService extends Service with Logging {
 
   def isRegistered: String
 
+  def decommission: String = throw new UnsupportedOperationException()
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala:70]: Yea, I think we can support this later?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:568]: Hmm for http response.. so add this delay, we can set it to other value, any idea?

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture @pan3793 @RexXiong  @cfmcgrady Any more suggestion?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala (hunk 7 above)
Comment (RexXiong) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:568]: 30s is too long, Maybe 10s

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 9]
```diff
@@ -592,7 +607,37 @@ private[celeborn] class Worker(
       logWarning(s"Waiting for all PartitionLocation release cost ${waitTime}ms, " +
         s"unreleased PartitionLocation: \n$partitionLocationInfo")
     }
-    stop(CelebornExitKind.WORKER_GRACEFUL_SHUTDOWN)
+  }
+
+  def decommissionWorker(): Unit = {
+    try {
+      masterClient.askSync(
+        ReportWorkerUnavailable(List(workerInfo).asJava),
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:615]: Currently shutdown workers would keep in master forever(if not come back), We need timeout these workers...

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala]: Done

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala (hunk 9 above)
Comment (pan3793) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:615]: I observed this issue too, I suppose it does not block this PR and should be addressed independently.

**Review by pan3793 — APPROVED** [review]

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala (hunk 9 above)
Comment (AngersZhuuuu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:615]: Yea, its not the issue related to this pr

**Comment by pan3793** [comment]
cc @turboFei, you may be interested in this feature.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 10]
```diff
@@ -561,6 +561,21 @@ private[celeborn] class Worker(
     sb.toString()
   }
 
+  override def decommission: String = {
+    exitKind = CelebornExitKind.WORKER_DECOMMISSION
+    new Thread() {
```
Comment (waitinfuture) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:566]: What's the purpose of this thread?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala (hunk 7 above)
Comment (pan3793) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala]: @waitinfuture see here

**Pull request closed without merging**